Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,999 @@
+theorem IsPurelyInseparable.bijective_algebraMap_of_isSeparable
+    [IsPurelyInseparable F E] [IsSeparable F E] : Function.Bijective (algebraMap F E) :=
+  ⟨(algebraMap F E).injective, surjective_algebraMap_of_isSeparable F E⟩
+
+variable {F E} in
+/-- If an intermediate field of `E / F` is both purely inseparable and separable, then it is equal
+to `F`. -/
+theorem IntermediateField.eq_bot_of_isPurelyInseparable_of_isSeparable (L : IntermediateField F E)
+    [IsPurelyInseparable F L] [IsSeparable F L] : L = ⊥ := bot_unique fun x hx ↦ by
+  obtain ⟨y, hy⟩ := IsPurelyInseparable.surjective_algebraMap_of_isSeparable F L ⟨x, hx⟩
+  exact ⟨y, congr_arg (algebraMap L E) hy⟩
+
+/-- If `E / F` is purely inseparable, then the separable closure of `F` in `E` is
+equal to `F`. -/
+theorem separableClosure.eq_bot_of_isPurelyInseparable [IsPurelyInseparable F E] :
+    separableClosure F E = ⊥ :=
+  bot_unique fun x h ↦ IsPurelyInseparable.inseparable F x (mem_separableClosure_iff.1 h)
+
+variable {F E} in
+/-- If `E / F` is an algebraic extension, then the separable closure of `F` in `E` is
+equal to `F` if and only if `E / F` is purely inseparable. -/
+theorem separableClosure.eq_bot_iff (halg : Algebra.IsAlgebraic F E) :
+    separableClosure F E = ⊥ ↔ IsPurelyInseparable F E :=
+  ⟨fun h ↦ isPurelyInseparable_iff.2 fun x ↦ have hx := (halg x).isIntegral; ⟨hx, fun hs ↦ by
+    simpa only [h] using mem_separableClosure_iff.2 hs⟩,
+      fun _ ↦ eq_bot_of_isPurelyInseparable F E⟩
+
+instance isPurelyInseparable_self : IsPurelyInseparable F F :=
+  ⟨fun _ ↦ isIntegral_algebraMap, fun x _ ↦ ⟨x, rfl⟩⟩
+
+variable {E}
+
+/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
+if and only if for every element `x` of `E`, there exists a natural number `n` such that
+`x ^ (q ^ n)` is contained in `F`. -/
+theorem isPurelyInseparable_iff_pow_mem (q : ℕ) [ExpChar F q] :
+    IsPurelyInseparable F E ↔ ∀ x : E, ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range := by
+  rw [isPurelyInseparable_iff]
+  refine ⟨fun h x ↦ ?_, fun h x ↦ ?_⟩
+  · obtain ⟨g, h1, n, h2⟩ := (minpoly.irreducible (h x).1).hasSeparableContraction q
+    exact ⟨n, (h _).2 <| h1.of_dvd <| minpoly.dvd F _ <| by
+      simpa only [expand_aeval, minpoly.aeval] using congr_arg (aeval x) h2⟩
+  have hdeg := (minpoly.natSepDegree_eq_one_iff_pow_mem q).2 (h x)
+  have halg : IsIntegral F x := by_contra fun h' ↦ by
+    simp only [minpoly.eq_zero h', natSepDegree_zero, zero_ne_one] at hdeg
+  refine ⟨halg, fun hsep ↦ ?_⟩
+  rw [hsep.natSepDegree_eq_natDegree, ← adjoin.finrank halg,
+    IntermediateField.finrank_eq_one_iff] at hdeg
+  simpa only [hdeg] using mem_adjoin_simple_self F x
+
+theorem IsPurelyInseparable.pow_mem (q : ℕ) [ExpChar F q] [IsPurelyInseparable F E] (x : E) :
+    ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range :=
+  (isPurelyInseparable_iff_pow_mem F q).1 (by assumption) x
```

Comment:
`‹_›` is interchangeable with `by assumption`, I believe. If you like it please replace everywhere.
```suggestion
  (isPurelyInseparable_iff_pow_mem F q).1 ‹_› x
```

Reply:
Thanks. I just can't remember how to type `‹` 😂

Reply:
It's `\f<`. Not as hard as ⟮, right? 😄